pictorus-stm32/src/pwm_protocol.rs

```diff
@@ -54,57 +53,196 @@ impl<T: timer::GeneralInstance4Channel> Pwm for PwmWrapper<'_, T> {
     }
 
     /// Sets the duty cycle from 0 to 1
-    fn set_duty(&mut self, _channel: Self::Channel, duty: Self::Duty) {
+    fn set_duty(&mut self, channel: Self::Channel, duty: Self::Duty) {
         let max_duty = self.simple_pwm.get_max_duty();
         let clamped_dc = duty.clamp(0.0, 1.0) as f32;
         let duty_final_u32 = clamped_dc.mul(max_duty as f32) as u32;
-        self.simple_pwm.set_duty(Channel::Ch1, duty_final_u32);
+        self.simple_pwm.set_duty(channel, duty_final_u32);
     }
 
     fn set_period<P>(&mut self, period: P)
     where
         P: Into<Self::Time>,
     {
-        // save current duty cycle period is in seconds for use later
-        let dc = self.get_duty(());
+        // save current duty cycle to use after frequency change
+        let (dc1, dc2, dc3, dc4) = self.get_duty_cycle_all();
+
         // Disable to make changes to the frequency
-        self.simple_pwm.disable(Channel::Ch1);
+        self.disable_all();
+
         let freq = 1.0 / period.into();
         // Note: the hz function takes a u32 value and set_frequency asserts if freq == 0, the minimum
         // PWM frequency must be an integer of 1 or greater.
         self.simple_pwm.set_frequency(hz(freq as u32));
+
         // Embassy set frequency requires a duty cycle update, since the max duty cycle changes
-        self.set_duty((), dc);
-        self.simple_pwm.enable(Channel::Ch1);
+        self.set_duty_cycle_all((dc1, dc2, dc3, dc4));
+
+        self.enable_all();
+    }
+}
+
+impl<T: timer::GeneralInstance4Channel> PwmWrapper<'_, T> {
+    fn disable_all(&mut self) {
+        self.disable_ch1();
+        self.disable_ch2();
+        self.disable_ch3();
+        self.disable_ch4();
+    }
+
+    fn enable_all(&mut self) {
+        self.enable_ch1();
+        self.enable_ch2();
+        self.enable_ch3();
+        self.enable_ch4();
+    }
+
+    fn get_duty_cycle_all(&self) -> (f64, f64, f64, f64) {
+        (
+            self.get_duty_ch1(),
+            self.get_duty_ch2(),
+            self.get_duty_ch3(),
+            self.get_duty_ch4(),
+        )
+    }
+
+    fn set_duty_cycle_all(&mut self, duty_cycle: (f64, f64, f64, f64)) {
+        self.set_duty_ch1(duty_cycle.0);
+        self.set_duty_ch2(duty_cycle.1);
+        self.set_duty_ch3(duty_cycle.2);
+        self.set_duty_ch4(duty_cycle.3);
+    }
+
+    fn enable_ch1(&mut self) {
+        if self.ch1.is_some() {
+            self.simple_pwm.enable(self.ch1.unwrap());
+        }
```

Comment:
Probably cleaner to do something like:
```
if let Some(ch) = self.ch1 {
  self.simple_pwm.enable(ch);
}
```

That way you don't need an unwrap

Reply:
Also wondering if rather than having explicit `<thing>_ch1`, `<thing>_ch2` methods it would be better to have a single:
```rust
fn enable_channel(&mut self, &channel: &Option<Channel>) {
    if let Some(ch) = channel {
        self.simple_pwm.enable(ch);
   }
}
```

Which you could call as:
```rust 
self.enable_channel(&self.ch1);
self.enable_channel(&self.ch2);
// etc.
```

That way you have less boilerplate/redundant methods